File: runtime/src/main/java/com/siimkinks/sqlitemagic/RawSelect.java
package com.siimkinks.sqlitemagic;

import android.arch.persistence.db.SupportSQLiteDatabase;
import android.database.Cursor;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.Size;

import java.util.Collection;

import static com.siimkinks.sqlitemagic.CompiledSelectImpl.createQueryObservable;
import static com.siimkinks.sqlitemagic.internal.ContainerHelpers.EMPTY_STRINGS;
import static java.lang.System.nanoTime;
import static java.util.concurrent.TimeUnit.NANOSECONDS;

/**
 * Builder for raw SQL SELECT statement.
 */
public final class RawSelect extends RawSelectNode<RawSelect, CompiledRawSelect> {
  RawSelect(@NonNull String sql) {
    super(new Builder(sql));
  }

  /**
   * Define tables that are involved in this SQL SELECT statement.
   * <p>
   * These tables are used in rx operations where subscribers will receive
   * an immediate notification for initial data as well as subsequent
   * notifications for when the supplied {@code table}'s data changes.
   *
   * @param tables Tables to select from. This param must be one of annotation processor
   *               generated table objects that corresponds to table in a database
   * @return A builder for raw SQL SELECT statement
   */
  @CheckResult
  public From from(@NonNull @Size(min = 1) Table<?>... tables) {
    final int len = tables.length;
    final String[] observedTables = new String[len];
    for (int i = 0; i < len; i++) {
      observedTables[i] = tables[i].nameInQuery;
    }
    return new From(this, observedTables);
  }

  /**
   * Define tables that are involved in this SQL SELECT statement.
   * <p>
   * These tables are used in rx operations where subscribers will receive
   * an immediate notification for initial data as well as subsequent
   * notifications for when the supplied {@code table}'s data changes.
   *
   * @param tables Tables to select from. This param must be one of annotation processor
   *               generated table objects that corresponds to table in a database
   * @return A builder for raw SQL SELECT statement
   */
  @CheckResult
  public <T extends Table<?>> From from(@NonNull @Size(min = 1) Collection<T> tables) {
    final int len = tables.size();
    final String[] observedTables = new String[len];
    int i = 0;
    for (Table<?> table : tables) {
      observedTables[i] = table.nameInQuery;
      i++;
    }
    return new From(this, observedTables);
  }

  /**
   * Builder for raw SQL SELECT statement.
   */
  public static final class From extends RawSelectNode<From, CompiledObservableRawSelect> {
    From(@NonNull RawSelect select, @NonNull String[] observedTables) {
      super(select.rawSelectBuilder);
      rawSelectBuilder.observedTables = observedTables;
    }

    /**
     * Create an observable which will notify subscribers with a {@linkplain Query query} for
     * execution.
     * <p>
     * Subscribers will receive an immediate notification for initial data as well as subsequent
     * notifications for when the supplied {@code table}'s data changes through the SqliteMagic
     * provided model operations. Unsubscribe when you no longer want updates to a query.
     * <p>
     * Since database triggers are inherently asynchronous, items emitted from the returned
     * observable use the {@link io.reactivex.Scheduler} supplied to
     * {@link com.siimkinks.sqlitemagic.SqliteMagic.DatabaseSetupBuilder#scheduleRxQueriesOn}. For
     * consistency, the immediate notification sent on subscribe also uses this scheduler. As such,
     * calling {@link io.reactivex.Observable#subscribeOn subscribeOn} on the returned observable has no
     * effect.
     * <p>
     * Note: To skip the immediate notification and only receive subsequent notifications when data
     * has changed call {@code skip(1)} on the returned observable.
     * <p>
     * One might want to explore the returned type methods for convenience query related
     * operators.
     * <p>
     * <b>Warning:</b> this method does not perform the query! Only by subscribing to the returned
     * {@link io.reactivex.Observable} will the operation occur.
     */
    @NonNull
    @CheckResult
    public SingleItemQueryObservable<Cursor> observe() {
      return new CompiledRawSelectImpl(rawSelectBuilder).observe();
    }
  }

  static final class Builder {
    @NonNull
    final String sql;
    @Nullable
    String[] args;
    @NonNull
    String[] observedTables = EMPTY_STRINGS;
    @NonNull
    DbConnectionImpl dbConnection = SqliteMagic.getDefaultDbConnection();

    Builder(@NonNull String sql) {
      this.sql = sql;
    }
  }

  static final class CompiledRawSelectImpl extends Query.DatabaseQuery<Cursor, Cursor> implements CompiledObservableRawSelect {
    @NonNull
    final String sql;
    @Nullable
    final String[] args;
    @NonNull
    final String[] observedTables;

    CompiledRawSelectImpl(@NonNull Builder builder) {
      super(builder.dbConnection, null);
      this.sql = builder.sql;
      this.args = builder.args;
      this.observedTables = builder.observedTables;
    }

    @NonNull
    @Override
    Cursor rawQuery(boolean inStream) {
      super.rawQuery(inStream);
      final SupportSQLiteDatabase db = dbConnection.getReadableDatabase();
      final long startNanos = nanoTime();
      final Cursor cursor = db.query(sql, args);
      if (SqliteMagic.LOGGING_ENABLED) {
        final long queryTimeInMillis = NANOSECONDS.toMillis(nanoTime() - startNanos);
        LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args);
      }
      return FastCursor.tryCreate(cursor);
    }

    @Override
    Cursor map(@NonNull Cursor cursor) {
      return cursor;
    }

    @NonNull
    @Override
    public Cursor execute() {
      return rawQuery(false);
    }

    @NonNull
    @Override
    public SingleItemQueryObservable<Cursor> observe() {
      return new SingleItemQueryObservable<>(createQueryObservable(observedTables, this));
    }

    @Override
    public String toString() {
      return "[RAW; sql=" + sql + "]";
    }
  }
}

File: runtime/src/main/kotlin/com/siimkinks/sqlitemagic/RawSelect.kt
package com.siimkinks.sqlitemagic

import android.database.Cursor
import androidx.annotation.CheckResult
import androidx.annotation.Size
import com.siimkinks.sqlitemagic.CompiledSelectImpl.Companion.createQueryObservable
import com.siimkinks.sqlitemagic.internal.ContainerHelpers.EMPTY_STRINGS
import java.util.concurrent.TimeUnit.NANOSECONDS

/** Builder for raw SQL SELECT statement. */
class RawSelect internal constructor(
  sql: String
) : RawSelectNode<RawSelect, CompiledRawSelect>(Builder(sql)) {
  /**
   * Define tables that are involved in this SQL SELECT statement.
   *
   * These tables are used in rx operations where subscribers will receive an immediate notification for initial data
   * as well as subsequent notifications for when the supplied `table`'s data changes.
   *
   * @param tables Tables to select from. This param must be one of annotation processor generated table objects that
   * corresponds to table in a database
   * @return A builder for raw SQL SELECT statement
   */
  @CheckResult
  fun from(@Size(min = 1) vararg tables: Table<*>) = From(
    select = this,
    observedTables = Array(tables.size) { index ->
      tables[index].nameInQuery
    }
  )

  /**
   * Define tables that are involved in this SQL SELECT statement.
   *
   * These tables are used in rx operations where subscribers will receive an immediate notification for initial data
   * as well as subsequent notifications for when the supplied `table`'s data changes.
   *
   * @param tables Tables to select from. This param must be one of annotation processor generated table objects that
   * corresponds to table in a database
   * @return A builder for raw SQL SELECT statement
   */
  @CheckResult
  fun <T : Table<*>> from(@Size(min = 1) tables: Collection<T>) = From(
    select = this,
    observedTables = tables
      .map { it.nameInQuery }
      .toTypedArray()
  )

  /** Builder for raw SQL SELECT statement. */
  class From internal constructor(
    select: RawSelect,
    observedTables: Array<String>
  ) : RawSelectNode<From, CompiledObservableRawSelect>(select.rawSelectBuilder) {
    init {
      rawSelectBuilder.observedTables = observedTables
    }

    /**
     * Create an observable which will notify subscribers with a [Query] for execution.
     *
     * Subscribers will receive an immediate notification for initial data as well as subsequent notifications for when
     * the supplied `table`'s data changes through the SqliteMagic provided model operations. Unsubscribe when you no
     * longer want updates to a query.
     *
     * Since database triggers are inherently asynchronous, items emitted from the returned observable use the
     * [io.reactivex.Scheduler] supplied to [SqliteMagic.DatabaseSetupBuilder.scheduleRxQueriesOn]. For consistency,
     * the immediate notification sent on subscribe also uses this scheduler. As such, calling
     * [io.reactivex.Observable.subscribeOn] on the returned observable has no effect.
     *
     * Note: To skip the immediate notification and only receive subsequent notifications when data has changed call
     * `skip(1)` on the returned observable.
     *
     * One might want to explore the returned type methods for convenience query related operators.
     *
     * **Warning:** this method does not perform the query! Only by subscribing to the returned
     * [io.reactivex.Observable] will the operation occur.
     */
    @CheckResult
    fun observe(): SingleItemQueryObservable<Cursor> =
      CompiledRawSelectImpl(builder = rawSelectBuilder)
        .observe()
  }

  internal class Builder internal constructor(
    @JvmField
    val sql: String
  ) {
    @JvmField
    var args: Array<out String>? = null

    @JvmField
    var observedTables: Array<String> = EMPTY_STRINGS

    @JvmField
    var dbConnection: DbConnectionImpl? = null
  }

  internal class CompiledRawSelectImpl internal constructor(
    @JvmField
    val sql: String,
    @JvmField
    val args: Array<out String>?,
    dbConnection: DbConnectionImpl?,
    @JvmField
    val observedTables: Array<String>
  ) : Query.DatabaseQuery<Cursor, Cursor>(
    dbConnection = dbConnection,
    mapper = null
  ), CompiledObservableRawSelect {
    internal constructor(builder: Builder) : this(
      sql = builder.sql,
      args = builder.args,
      dbConnection = builder.dbConnection,
      observedTables = builder.observedTables
    )

    override fun rawQuery(
      inStream: Boolean,
      dbConnection: DbConnectionImpl
    ): Cursor {
      super.rawQuery(
        inStream = inStream,
        dbConnection = dbConnection
      )
      val db = dbConnection.readableDatabase
      val startNanos = System.nanoTime()
      val cursor = SqlUtil.query(db, sql, args)
      if (SqliteMagic.LOGGING_ENABLED) {
        val queryTimeInMillis = NANOSECONDS.toMillis(System.nanoTime() - startNanos)
        LogUtil.logQueryTime(queryTimeInMillis, observedTables, sql, args)
      }
      return FastCursor.tryCreate(cursor)
    }

    override fun map(
      cursor: Cursor?,
      dbConnection: DbConnectionImpl
    ) = cursor

    override fun execute(): Cursor = checkNotNull(rawQuery(inStream = false))

    override fun observe(): SingleItemQueryObservable<Cursor> = SingleItemQueryObservable(
      createQueryObservable(
        observedTables = observedTables,
        query = this
      )
    )

    override fun toString() = "[RAW; sql=$sql]"
  }
}
